src/flush.rs

```diff
@@ -39,41 +43,71 @@ where S: Body,
-                    true
-                } else {
-                    false
-                };
-                self.h2.send_data(buf, eos)?;
-                if eos {
+            match try_ready!(self.poll_body()) {
+                Some(Data(buf)) => {
+                    let eos = self.body.is_end_stream();
+
+                    self.h2.send_data(SendBuf::new(buf.into_buf()), eos)?;
+
+                    if eos {
+                        self.state = FlushState::Done;
+                        return Ok(Async::Ready(()));
+                    }
+                }
+                Some(Trailers(trailers)) => {
+                    self.h2.send_trailers(trailers)?;
+                    return Ok(Async::Ready(()));
+                }
+                None => {
+                    // If this is hit, then an EOS was not reached via the other
+                    // paths. So, we must send an empty data frame with EOS.
+                    self.h2.send_data(SendBuf::none(), true)?;
+
                     return Ok(Async::Ready(()));
-                } else if let Async::Ready(item) = second {
-                    first = item;
-                } else {
-                    return Ok(Async::NotReady);
                 }
-            } else if let Some(DataOrTrailers::Trailers(trailers)) = first {
-                self.h2.send_trailers(trailers)?;
-                return Ok(Async::Ready(()));
-            } else {
-                return Ok(Async::Ready(()));
             }
         }
     }
 
     /// Get the next message to write, either a data frame or trailers.
-    fn poll_body(&mut self) -> Poll<Option<DataOrTrailers<S::Data>>, h2::Error> {
+    fn poll_body(&mut self)
+        -> Poll<Option<DataOrTrailers<S::Data>>, h2::Error>
+    {
         loop {
             match self.state {
                 FlushState::Data => {
+                    // Before trying to poll the next chunk, we have to see if
+                    // the h2 connection has capacity. We do this by requesting
+                    // a single byte (since we don't know how big the next chunk
+                    // will be.
+                    self.h2.reserve_capacity(1);
+
+                    if self.h2.capacity() == 0 {
+                        // TODO: The loop should not be needed once
+                        // carllerche/h2#270 is fixed.
+                        loop {
+                            match try_ready!(self.h2.poll_capacity()) {
+                                Some(0) => {}
+                                Some(_) => break,
+                                None => {
+                                    debug!("connection closed early");
+                                    // The error shouldn't really matter at this
+                                    // point as the peer has disconnected, the
+                                    // error will be discarded anyway.
+                                    return Err(h2::Reason::INTERNAL_ERROR.into());
```

Comment:
It might be useful for a user to know the reason for the reset. Or is that in the `Err` variant from `poll_capacity`? What does `Ok(None)` mean, then?

Reply:
Yeah, I'm not entirely sure if this is ever hit... If the remote disconnects then it probably will be bubbled up as an error. I didn't really want to add a panic in that branch, but I could.